### ISSUE TITLE
#define Val_some

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#5154]
**Original bug ID:** 5154
**Reporter:** @ygrek
**Assigned to:** @damiendoligez
**Status:** assigned (set by @mshinwell on 2016-12-08T08:46:17Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Version:** 3.11.2
**Category:** runtime system and C interface
**Tags:** patch
**Monitored by:** @hcarty

## Bug description

OCaml C interface defines many macros to construct/access primitive ocaml values from C code. One thing that I often find missing and have to copy-paste myself in almost every binding is code to handle option type.

## Additional information

#define Val_none Val_int(0)

value Val_some(value v)
{
    CAMLparam1(v);
    CAMLlocal1(some);
    some = caml_alloc_small(1, 0);
    Field(some, 0) = v;
    CAMLreturn(some);
}

#define Some_val(v) Field(v,0)

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

@doligez I find the lack of this functionality to be a bit of a nuisance as well; what do you think?

**Comment by ygrek** [comment]
Val_some(activity)

**Comment by yallop** [comment]
@ygrek: would you consider submitting a PR?

**Comment by nojb** [comment]
I took the liberty of making a PR with his patch #9569

**Comment by MisterDA** [comment]
Just for reference, some projects had already defined some of the macros. For instance, see [ocaml-mccs](https://github.com/AltGr/ocaml-mccs/blob/0a8fc4aea7d9ab4098388b118540f347ab1dead1/src/mccs_stubs.cpp#L26). Although they share the same definition, some compilers raise warnings when redefining macros (e.g. MSVC 2019). In some setups, warnings are turned to errors and the build of a package fails (e.g. building Opam…).

I’m submitting patches to fix that, I think that checking `OCAML_VERSION` is the best way:

```c
#if OCAML_VERSION < 41200
#define Val_none Val_int(0)
#define Some_val(v)  Field(v, 0)
#endif
```

EDIT: changed `<= 41100` to `< 41200`.

**Comment by nojb** [comment]
Thanks for the heads-up @MisterDA! Yes, checking `OCAML_VERSION` seems a good fix.

**Comment by MisterDA** [comment]
But which version? If the commit won’t be included before 4.12, then `< 41200` is a better choice in case there’s a 4.11.x.

**Comment by nojb** [comment]
Indeed, the patch in question will only be included in 4.12, and not in the 4.11.x releases.
